src/main/java/com/louis/BinarySearchTree.java

```diff
@@ -0,0 +1,283 @@
+// CHECKSTYLE:OFF
+
+// CHECKSTYLE:ON
+
+package com.louis;
+
+import com.google.common.base.Preconditions;
+import java.util.NoSuchElementException;
+
+/**
+ * A binary search tree implementation.
+ * <p>
+ * Binary search trees (BSTs) are derived from the associative array abstract data type. It is a symbol table with keys
+ * and values; null is excluded. The tree is not balanced. Ks are Comparable and the tree is sorted by them. Most
+ * operations have linear worst case complexity.
+ *
+ * @author Louis Ashton (louisashton@live.com)
+ */
+public class BinarySearchTree<K extends Comparable<K>, V> {
+
+    // The root of the binary search tree.
+    private Node root;
+
+    /**
+     * A node of a binary search tree.
+     * <p>
+     * Each node has a key and associated value. Pointers to the node's children are also stored.
+     */
+    private class Node {
```

Comment:
My personal preference is to avoid non static inner classes - having them static makes it easier to reason about how they are used.